### PR TITLE
add API endpoint to create OAuth2 Application.

### PR DESCRIPTION
**Pull request by ghost**
This PR adds support to create OAuth2 Applications through the API. The intention is to be able to automate the creation of application programmatically. 

Maybe it is  related to the issue #8764 and #9210

### CODE REVIEW

File: routers/api/v1/user/user.go [hunk 1]
```diff
@@ -166,3 +166,24 @@ func GetUserHeatmapData(ctx *context.APIContext) {
 	}
 	ctx.JSON(http.StatusOK, heatmap)
 }
+
+// CreateOauth2Application is the handler to create a new OAuth2 Application for the authenticated user
+func CreateOauth2Application(ctx *context.APIContext, data api.CreateOauthApplicationOption) {
```
Comment (adelowo) [inline routers/api/v1/user/user.go]: Please add swagger annotations for this handler then run `make generate-swagger`

**Review by adelowo — CHANGES_REQUESTED** [review]

File: modules/structs/user_app.go [hunk 2]
```diff
@@ -32,3 +32,19 @@ type AccessTokenList []*AccessToken
...
+type CreateOAuth2ApplicationOptions struct {
+	Name         string   `json:"name" binding:"Required"`
+	RedirectURIs []string `json:"redirect-uris" binding:"Required"`
+}
+
+// OAuth2Application represents an OAuth2 application.
+// swagger:response OAuth2Application
+type OAuth2Application struct {
+	Name         string   `json:"name"`
+	ClientID     string   `json:"client-id"`
```
Comment (adelowo) [inline modules/structs/user_app.go]: JSON tags should use underscore not a dash. Consistency with the rest of the api.

**Review by adelowo — CHANGES_REQUESTED** [review]

File: modules/structs/user_app.go (hunk 2 above)
Comment (ghost) [reply modules/structs/user_app.go]: thanks. I've updated it on the commit below. Also added the docs for the request body.

**Review by adelowo — APPROVED** [review]

**Review by adelowo — CHANGES_REQUESTED** [review]
Need to fix swagger though.

**Comment by IGZgustavomarin** [comment]
> Need to fix swagger though.

I've added CreateOAuth2ApplicationOptions to the swagger body annotation. The swagger generates ok with `make swagger-generate` but the pipeline fails on `make swagger-validate` with this error:

```
- "paths./user/applications/oauth2.post.parameters" must validate one and only one schema (oneOf). Found none valid
- paths./user/applications/oauth2.post.parameters.type in body is a forbidden property
- invalid definition for parameter body in body in operation "userCreateOAuth2Application"

```

Not sure how to fixed. Any help?

File: routers/api/v1/user/app.go [hunk 3]
```diff
@@ -135,3 +135,43 @@ func DeleteAccessToken(ctx *context.APIContext) {
...
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: body
+	//   in: body
+	//   required: true
+	//   schema:
+	//     "$ref": "#/definitions/CreateOAuth2ApplicationOptions"
+	// responses:
+	//   "200":
```
Comment (6543) [inline routers/api/v1/user/app.go:153]:
```suggestion
	//   "201":
```

File: routers/api/v1/user/app.go [hunk 4]
```diff
@@ -135,3 +135,43 @@ func DeleteAccessToken(ctx *context.APIContext) {
...
+		return
+	}
+	secret, err := app.GenerateClientSecret()
+	if err != nil {
+		ctx.Error(http.StatusBadRequest, "", "error creating application secret")
+		return
+	}
+	app.ClientSecret = secret
+
+	ctx.JSON(http.StatusOK, api.OAuth2Application{
```
Comment (6543) [inline routers/api/v1/user/app.go:171]:
```suggestion
	ctx.JSON(http.StatusCreated, api.OAuth2Application{
```

**Review by 6543 — CHANGES_REQUESTED** [review]
please read https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#api-v1

**Comment by 6543** [comment]
@IGZgustavomarin can you also add a GET method witch return a list of OAuth2 apps (without secret) and a Delete method?

**Comment by IGZgustavomarin** [comment]
> @IGZgustavomarin can you also add a GET method witch return a list of OAuth2 apps (without secret) and a Delete method?

Yes, I'm doing this, but also am moving my changes to a branch so I can keep more easily up master branch changes.

**Pull request closed without merging**